13604 local testing of GitHub actions

radoslawkrzemien opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Testing framework for running local tests of GitHub Actions workflows to identify issues before pushing them into the repository and testing there

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/13604 
$ https://github.com/Expensify/App/issues/13604#issuecomment-1405005943


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Run `npm run workflow-test`
2. Verify test run results

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
3. Upload an image via copy paste
4. Verify a modal appears displaying a preview of that image
--->
1. Run `npm run workflow-test`
2. Verify test run results

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

radoslawkrzemien committed:
Add Mock-github folder to .gitignore

Updated the `.gitignore` file with the temporary folder created by
the `Mock-github` package, in case the cleanup was not performed for
any reason

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add new dependencies

Added new packages, most importantly `Mock-github` for creating a
local repository for testing (to have better control over which
files are accessed by the locally run workflows) and `Act-js` for
programatically triggering the workflows with Act (with features
added on top of it, like mocking API calls, mocking steps, etc.)

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add configs

Added `jest` and `ts` config files in a new folder `workflow_tests`
for the testing of GitHub Actions workflows

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add first proper test stub [WiP]

Added the stub of the first proper test, with the more-or-less defined structure, but yet without proper step mocks and result assertions

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Refactor test configuration [WiP]

Updated the mocks/test repo files configuration to be more
readable/maintainable. Also added `name` properties to `preDeploy`
workflow steps, since using `name` is the easiest and most stable way
of referencing the steps for mocking

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix steps mocks [WiP]

Updated the step mocks to allow for producing outputs that steer the
further workflow execution. Also added run params to the act call

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add output assertions [WiP]

Added the option to assert the output of workflow run and compare it
to a set of expected values

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests for failing lint/test jobs [WiP]

Added the tests to check what happens when either of `lint` or
`test` jobs fails, namely if `confirmPassingBuild` job gets executed
and all jobs depending on `lint` and `test` being successful do not

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Refactor test structure [WiP]

The test file is getting way too big, I will have to split it into
several files, but for now did some refactoring for maintainability

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Clean up test file [WiP]

Extracted step mocks, assertion helper methods and utils methods to their separate files to clean up the already cluttered test file

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add new tests [WiP]

Added tests for new contributor message job

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Migrate from .ts to .js [WiP]

Changed all typescript code into javascript code and removed ts-related dependencies

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add method for getting mock steps [WiP]

Added a helper method to utils, that helps with creating step mocks - easier than mocking them by hand, allows reusability and reduces the chance of a typo sneaking in

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add method for getting step assertions [WiP]

Added a helper method to utils, that helps with creating mock step assertions - easier than typing them by hand, allows reusability and reduces the chance of a typo sneaking in

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Remove tsconfig [WiP]

Forgot to previously remove tsconfig.json

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests for chooseDeployActions/skipDeploy/createNewVersion/updateStaging [WiP]

Added the tests checking for which steps get executed based on given set of conditions

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests for push to a different branch [WiP]

Added the test checking if the workflow does not get executed if the code is pushed to a different branch than main

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add test for events different from push [WiP]

Added the test checking if the workflow does not get executed if the triggering event is different from `push`

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add documentation skeleton [WiP]

Started working on the tests documentation

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add file structure to documentation [WiP]

Added a new section to documentation with structure of the workflow testing framework

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add utils description [WiP]

Added a new section to documentation with the description and example usage of utils helper methods

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add breakdown of the test file structure [WiP]

Added a new section to documentation with the detail breakdown of the structure of typical test file

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add minor tweaks [WiP]

Self-explanatory

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add test for Slack announcement [WiP]

Added the test checking if the failed workflow gets announced on Slack

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add setup section [WiP]

Added new section to readme, about setting up the environment for running tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add deploy tests [WiP]

Added tests for the deploy.yml workflow and fixed the setActParams method

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Clean up preDeploy.test.js [WiP]

Self-explanatory

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
I have read the CLA Document and I hereby sign the CLA

radoslawkrzemien committed:
Tweak preDeploy config[WiP]

Updated event options and local repo pushed branches

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add test for different event

Added test checking if deploy workflow does not execute on a different event than push

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add first platformDeploy test

Added test first working test for platformDeploy workflow, along with all the mocks and assertions

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add method to overwrite the runner type

Some runner types are not supported by act. To run the jobs using them, even with mock steps to check the execution flow and outputs, I've added a helper method to overwrite the runner type

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update readme

Added a description and examples regarding the new method for overwriting the runner types

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update test

Utilised the new method to run all jobs within the `platformDeploy` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add dependency

Added yaml to the dependencies, even though it's still required by act-js

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Extend Act class

Had to extend the `Act` class from `Act-js` because there was no apparent way to change the `github.actor` property on which many workflows depend

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update tests

Updated the tests with extended Act class and assertions with single step checks

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update tests

Updated previously failing tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Restructure utils

Created utils folder, moved ExtendedAct into a separate file

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Pre-generate test

Added a script to automatically pre-generate a test file, mocks file, and an assertions file for a given workflow to speed up tests development

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add script to package.json

Added the new script to package.json under `workflow-test:generate`

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update readme

Updated the workflow_tests readme with mention of the new script

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix imports

Fixed imports after the refactor of utils

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update pre-generation script

Removed replacing spaces in generated mock/assertion names

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update setting event options

Updated the way event data is set for `Act` instance, since `Act` seems to not be consistent in how it expects the data to be provided based on the type of event. This change seems to cover all possibilities. Plus renamed some variables to be compliant with JS convention

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add step names

Added missing names to steps in the `lockDeploys.yml` workflow for test files pre-generation

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests, mocks, and assertions for `lockDeploys.yml` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update `finishReleaseCycle.yml`

Updated workflow file for pre-generation of test files

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add more tests

Added tests for the `finishReleaseCycle.yml` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add npm script

Added the `workflow-test` script to `package.json` to simplify triggering the workflow tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `authorChecklist` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update readme

Added a mention about `Act`'s limitation to the workflow tests Readme

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Tweak pre-generation script

Made the job's name capitalised in the assertion method names

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added test for the `cherryPick.yml` workflow, plus edited the utils to allow for setting the event inputs

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added test for the `cla.yml` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added test for the `createNewVersion.yml` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Refactor tests

Slightly refactored existing tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `deployBlocker` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `lint` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `reviewerChecklist` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `test` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `test` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `validateGithubActions` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `warnCPLabel` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `verifySignedCommits` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `verifyPodfile` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Add tests

Added tests for `testBuild` workflow

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed `cherryPick` tests that started failing after a recent refactor

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update readme

Updated FAQ

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix test

Fix failing `deploy` tests after a recent refactor

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update package.json

Removed redundant `-t` filter from command

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into 13604-local-testing-of-github-actions

# Conflicts:
#	.github/workflows/preDeploy.yml
#	package-lock.json
#	package.json

radoslawkrzemien committed:
Update package-lock.json

Update after merging main

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix tests

Fixed `preDeploy` tests that broke after merging in the source repos main branch

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
Also, have you followed the `Setup` section in the `README`? Namely the step where you set the `ACT_BINARY` environment variable?

dangrous commented:
So the main error is just receiving empty arrays when workflows should run. It looks like the only passes are for "workflow does not execute" or "does not run the lint" etc. etc., so nothing is ever running. For example:
```
 ● test workflow deploy › push as user › to production - nothing triggered

    expect(received).toEqual(expected) // deep equality

    Expected: ArrayContaining [{"name": "Main Get merged pull request", "output": "[MOCK] [VALIDATE] Getting merged PR, github_token=***", "status": 0}]
    Received: []

      20 |     for (const expectedStep of steps) {
      21 |         if (didExecute) {
    > 22 |             expect(workflowResult).toEqual(expect.arrayContaining([expectedStep]));
         |                                    ^
      23 |         } else {
      24 |             expect(workflowResult).not.toEqual(expect.arrayContaining([expectedStep]));
      25 |         }

      at Object.toEqual [as assertValidateJobExecuted] (assertions/deployAssertions.js:22:36)
      at Object.assertValidateJobExecuted (deploy.test.js:236:24)
```
I didn't see the README before, but followed the steps now and still getting the same issue. When I try to run `act` on its own, it tells me:
```
ERRO[0000] failed to obtain container engine info: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
[Deploy code to staging or production/validate                        ] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Deploy code to staging or production/validate                        ]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=true
[Build and deploy android, desktop, iOS, and web clients/validateActor] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Build and deploy android, desktop, iOS, and web clients/validateActor]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=true
[Process new code merged to main/isExpensifyEmployee                  ] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Process new code merged to main/isExpensifyEmployee                  ]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=true
Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```
Do we need to do something with docker?

radoslawkrzemien commented:
Hm, I didn't have that issue. Let me look into it further and get back to you

radoslawkrzemien commented:
Actually no, I have the same problem regardless of directory. This is new, I run the tests successfully yesterday morning. Let me look into it some more

radoslawkrzemien committed:
Fix workflow file

Fixed previously broken `preDeploy` workflow file

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
Ok, in the last commit, while fixing a broken test I have unintentionally added the `steps` key to a workflow file twice which broke the workflow file instead 🤦 Should be better now

radoslawkrzemien commented:
Regarding docker - you need to have docker running locally for it to work as it `Act`'s prerequisite, per their docs - https://github.com/nektos/act#necessary-prerequisites-for-running-act

mczernek commented on workflow_tests/README.md:
```diff
@@ -0,0 +1,408 @@
+
+## Setup
+- Install dependencies from `package.json` file with `npm install`
+- Install `Act` with `brew install act` and follow the documentation on [first Act run](https://github.com/nektos/act#first-act-run)
```
I think we should link [necessary-prerequisites-for-running-act](https://github.com/nektos/act#necessary-prerequisites-for-running-act) as well

radoslawkrzemien replied:
Added another point in the `Setup` section about meeting the prerequisites and linking the `Act` docs

mczernek commented on workflow_tests/README.md:
```diff
@@ -0,0 +1,408 @@
+- Install `Act` with `brew install act` and follow the documentation on [first Act run](https://github.com/nektos/act#first-act-run)
+- Set the environment variable `ACT_BINARY` to the path to your `Act` executable (`which act` if you're not sure what the path is)
+- You should be ready to run the tests now with `npm test -- --config=workflow_tests/jest.config.js`
+- You can pre-generate new mocks/assertions/test files for a given workflow by running `npm run workflow-test:generate <workflow_file>`
```
I think we should also have `Running` section, maybe just simply `npm run workflow-test` in there, but still :)

radoslawkrzemien replied:
Added a new section with information on how to run the tests

radoslawkrzemien committed:
Move dependencies

Moved the dependencies for testing workflows to dev-dependencies

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Move repo files

Moved the test repo files definition to the utils file

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented on workflow_tests/deploy.test.js:
```diff
@@ -0,0 +1,294 @@
+jest.setTimeout(60 * 1000);
+let mockGithub;
+const FILES_TO_COPY_INTO_TEST_REPO = [
+    {
```
Moved them to `utils.js`

radoslawkrzemien committed:
Rename functions

Renamed two of the utils functions from `get...` to `create...`

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Rename references

Renamed the references of the previously renamed functions

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented on workflow_tests/utils/utils.js:
```diff
@@ -0,0 +1,160 @@
+    return updated_act;
+}
+
+function getMockStep(
```
Renamed `getMockStep` to `createMockStep` and `getStepAssertion` to `createStepAssertion`

radoslawkrzemien committed:
Update readme

Added the link to `Act` documentation on necessary prerequisites (like Docker)

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Update readme

Added a new section on running the tests

See: https://github.com/Expensify/App/issues/13604

dangrous commented:
Still struggling with getting this set up. I've installed Docker and have it running, and it's successfully interacting with act. However, still getting all the same failures. Running `act` gives me a little more visibility on what's happening, but not yet how to fix it. 

Primarily, it's showing `  ❗  ::error::Input required and not supplied: GITHUB_TOKEN` which is weird because I ran it specifically entering the GITHUB_TOKEN (`act --container-architecture linux/amd64 -s GITHUB_TOKEN` and then entered it as secure input) but it seems like that's not persisting. Also, even so, I'm not sure how to include a token in `npm run workflow-test`. 

Also, I had to add `GITHUB_REPOSITORY=dangrous/testing` to my env as well (along with ACT_BINARY) to get that far; otherwise I get a lot of `unable to get git repo: branch config: invalid merge` errors when running `act`. Do I need to use the actual `Expensify/App` repo? If so, are there any potential things this would create/delete/mess up/etc. if I point it there?

Sorry for all the questions, just want to make sure this is working before reviewing the code and the prereqs/setup are a bit unclear as of now.

radoslawkrzemien commented:
@dangrous I am looking into that.
You don't need to set the `GITHUB_REPOSITORY` variable, the tests create a local copy of the repository before running, run in the context of that local copy and then destroy it after - specifically so they can't interact with the real repository and mess something up.
Straight up running `act` executes the workflows (not the tests) in the context of the main repo so it can tell you if something is wrong with your `act` configuration, but it is not the same as running the tests themselves.
What I did to successfully run the tests on a fresh repo:
- had `act` and `docker` installed
- created a new directory
- cloned the repo
- checked out the PR branch
- `npm install`
- set the `ACT_BINARY`
- `npm run workflow-test`
- works
@mczernek afaik you also managed to run the tests without issue, can you remember if you had to do anything more than is described here or in the `README`?

mczernek commented:
AFAIR I've done just the steps you've described and it went well. I'll give it another go on another machine.

dangrous commented:
Macbook pro with the M1 Max chip.

Although I'm now trying it on another older macbook with an intel chip, so I can report back on that.

@AndrewGable or @sobitneupane were you able to get this working?

radoslawkrzemien commented:
Are the errors still like 
```
    Expected: ArrayContaining [{"name": "Main Get merged pull request", "output": "[MOCK] [VALIDATE] Getting merged PR, github_token=***", "status": 0}]
    Received: []
```
?
If so, have you pulled the most recent commits fixing the broken workflow file?

sobitneupane commented:
Getting [the same errors](https://github.com/Expensify/App/pull/15560#issuecomment-1490237489) as @dangrous on `npm run workflow-test`

mczernek commented:
@sobitneupane @dangrous Seems like act fails to connect to docker.

Let's try two things:
1. edit `~/.docker/config.json` by replacing `credsStore` with `credStore`
2. `sudo ln -s ~/.docker/run/docker.sock /var/run/docker.sock`

Then remove `repo` dir and rerun tests.

dangrous commented:
Hm yeah I think you're right about it not connecting - when I run just `act` in `App` I see activity in Docker Desktop implying that it's doing, er, something. But, none of that information updates when I run `npm run workflow-test`.

Unfortunately, that still happens after running the commands you [suggested](https://github.com/Expensify/App/pull/15560#issuecomment-1505113374). 

And yes, same empty array errors, and have pulled and am up to date.

Just in case it's helpful, here's the most successful run I can get with just `act`. I know this isn't what we're looking for, but maybe those two errors at the end (or the options I need to give it) could be directional...

```
App % (13604-local-testing-of-github-actions) act --container-architecture linux/amd64 -s GITHUB_TOKEN="***"
[Build and deploy android, desktop, iOS, and web clients/validateActor] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Deploy code to staging or production/validate                        ] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Process new code merged to main/isExpensifyEmployee                  ] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[Build and deploy android, desktop, iOS, and web clients/validateActor]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 username= forcePull=true
[Process new code merged to main/isExpensifyEmployee                  ]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 username= forcePull=true
[Deploy code to staging or production/validate                        ]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 username= forcePull=true
[Process new code merged to main/isExpensifyEmployee                  ]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Deploy code to staging or production/validate                        ]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Process new code merged to main/isExpensifyEmployee                  ]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Deploy code to staging or production/validate                        ]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Build and deploy android, desktop, iOS, and web clients/validateActor]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Build and deploy android, desktop, iOS, and web clients/validateActor]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Deploy code to staging or production/validate                        ]   ☁  git clone 'https://github.com/roryabraham/action-get-merged-pull-request' # ref=7a7a194f6ff8f3eef58c822083695a97314ebec1
[Process new code merged to main/isExpensifyEmployee                  ]   ☁  git clone 'https://github.com/roryabraham/action-get-merged-pull-request' # ref=7a7a194f6ff8f3eef58c822083695a97314ebec1
[Build and deploy android, desktop, iOS, and web clients/validateActor]   ☁  git clone 'https://github.com/tspascoal/get-user-teams-membership' # ref=baf2e6adf4c3b897bd65a7e3184305c165aec872
[Deploy code to staging or production/validate                        ] 🏁  Job succeeded
[Process new code merged to main/isExpensifyEmployee                  ] 🏁  Job succeeded
[Process new code merged to main/isExpensifyEmployee                  ] Unable to interpolate expression 'format('{0}', fromJSON(steps.checkActor.outputs.isTeamMember))': Cannot parse non-string type invalid as JSON
[Build and deploy android, desktop, iOS, and web clients/validateActor] 🏁  Job succeeded
[Build and deploy android, desktop, iOS, and web clients/validateActor] Unable to interpolate expression 'format('{0}', fromJSON(steps.isUserDeployer.outputs.isTeamMember) || github.actor == 'OSBotify')': Cannot parse non-string type invalid as JSON
Error: Unable to determine how to run job:lint step:Run lint workflow
```

dangrous commented:
Follow up - would it be worth scheduling a call at some point to screenshare and try to figure this all out? Might be faster than the back and forth on GH. Let me know!

dangrous commented:
Hi all! Checking in to see if there's any progress on this? I'm happy to keep testing if we've got some new ideas.

radoslawkrzemien commented:
I'm working on a different project now, but I will try to reproduce the problem on my side and find a solution whenever I have the time. Will get back if I find anything

radoslawkrzemien committed:
Add logging raw output to file

Added logging the raw output from `act` into files under `workflow-tests/logs/<workflow_name>.log`. Raw output should have more information in case there are some errors instead of just returning an empty array like `act-js` does

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
@dangrous I've added printing the raw output from `act` into a file, this should help identify the issue, as `act-js` parses the output into an array and if there are any errors it just returns `[]` which is not helpful.
Please run any of the previously failing tests (preferably with either `-i` or `-t` flag to save time) and check the contents of the `workflow-tests/logs/*.log` file(s) generated. If the tests still fail, please paste the contents here so we can have a look and think about the solution

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into 13604-local-testing-of-github-actions

# Conflicts:
#	.github/workflows/testBuild.yml
#	package-lock.json

radoslawkrzemien committed:
Update fork

Merged changes from main repo

See: https://github.com/Expensify/App/issues/13604

dangrous commented:
Okay on first run of this - I'm not getting any logfiles generated, but on the bright side the in-console logs have changed. Now instead of an empty array for all the errors I now see:

<img width="1148" alt="image" src="https://user-images.githubusercontent.com/5487802/235187050-847f9553-b09b-469b-8621-5b56fce8987f.png">

I also tried running just `act` again on its own and got more loglines than usual so maybe something has changed (it feels like it might be getting farther in the process than it did before.

radoslawkrzemien committed:
Update workflow paths and fix test after merge

After merging the base repo main branch `act-js` got updated to version 2.0.8 and now requires a path to the workflow file instead of a path to workflows directory. Also, one of the workflows has been changed in the meantime so the test had to be adjusted

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
I have synced the fork with the main repo and updated the dependencies, which updated `act-js` to version 2.0.8 and it now requires a full path to the workflow file instead of a path to workflows directory when calling `runEvent()` - that's why you had `Could not locate workflows`. I updated the calls, so please try again and let me know what the result is

dangrous commented:
Okay, this is going much better! Tests all ran this time. Notes:
* One test failed:
```
 FAIL  workflow_tests/preDeploy.test.js (239.925 s)
  ● test workflow preDeploy › push to main - workflow executes

    error: remote origin already exists.

      at Object.action (../node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)
      at PluginStore.exec (../node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)
      at ../node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42
      at GitExecutorChain.handleTaskData (../node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)
      at GitExecutorChain.<anonymous> (../node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)
      at fulfilled (../node_modules/simple-git/dist/cjs/index.js:55:24)
          at runMicrotasks (<anonymous>)
```
* This takes an incredibly long time - 18+ minutes for the full set of tests. Is that expected?
* Not a huge deal, but I'm not sure the logfiles are working correctly. They fill up with logs but only the ones from the final test of each suite are saved (the others are deleted when new ones are written)

Now that it's actually running, I can start looking at the actual code itself. A couple questions about that:
* Is there a good reference for what we're hoping to have tested, to make sure we're covering all the cases we want, etc.? I didn't notice a list or anything in the issue.
* Is there a way you'd recommend going through this, things that are tricky that I should take extra care to review? It's pretty big, so it'll be good to have a plan of attack for the review.

Thanks!

AndrewGable reviewed (commented):
Wow this is a lot of code! Just a few comments I saw initially. I will need some more time to review the whole thing.

radoslawkrzemien commented:
@dangrous Regarding the long run times - it is expected to run that long - even with simple tests with mocked steps which only print stuff act still needs to spin up containers for each step, etc. which takes time.

I'll look into removing the test repo before the tests and see what can be done.

Yes, the log files are overwritten by each test so they only have the logs from the last one run in a suite. I used the built in feature of `act-js` for that, so changing it to appending instead of overwriting is possible but problematic. What else can be done is to have a separate log files for each test, which will create 124 log files, so it might be good to, for example, have each test suite create a subdirectory in the `logs/` directory and have the separate log files in there

The main points which I had in mind when approaching these tests were those:
- if the workflow uses a step defined by the js action - no point in testing that step because it can be tested separately
- if the workflow uses a step defined by another workflow - no point in testing that step because it can be tested separately
- steps performing actions on the repository like posting comments, etc. are hard to test behaviourally, but I guess it can be assumed that built-in GA actions perform as expected and they can be mocked
- steps that perform checks, set env variables, outputs, etc. or print stuff can be left as is
- the main thing that can be tested here is which steps get executed under which conditions and in which order, plus how are the variables and outputs being set - that's why most of the steps are mocked with predetermined output which is then validated to see if that step had been executed or not
- created several utils and helpers which should help with maintaining and expanding the tests further

As for "attack plan", I'd probably suggest to check whether the tests cover the usual scenarios that happen with your workflows, as you probably have a more in-depth knowledge about that than me - I never had a chance to actually interact with them during development, just worked with the workflow files and deduced their function. Also, apart from the tests themselves I guess the most important thing would be to check the utils and helpers

radoslawkrzemien commented on .github/workflows/cla.yml:
```diff
@@ -13,18 +13,20 @@ jobs:
             regex: '\s*recheck\s*'
         - name: CLA Assistant
-          if: ${{ steps.recheck.outputs.match != '' || steps.sign.outputs.match != '' }} || github.event_name == 'pull_request_target'
+          if: ${{ steps.recheck.outputs.match != '' || steps.sign.outputs.match != '' || github.event_name == 'pull_request_target' }}
```
If I remember correctly `github.event_name == 'pull_request_target'` being outside of the `${{}}` block never resolved to `true` when running tests so I assumed it was a bug

radoslawkrzemien commented on .github/workflows/preDeploy.yml:
```diff
@@ -2,37 +2,46 @@ name: Process new code merged to main
-      IS_AUTOMATED_PR: ${{ steps.isAutomatedPullRequest.outputs.IS_AUTOMATED_PR }}
-      SHOULD_DEPLOY: ${{ steps.shouldDeploy.outputs.SHOULD_DEPLOY }}
-      SHOULD_CP: ${{ steps.isStagingDeployLocked.outputs.IS_LOCKED && steps.hasCherryPickLabel.outputs.HAS_CP_LABEL }}
+      IS_AUTOMATED_PR: ${{ fromJSON(steps.isAutomatedPullRequest.outputs.IS_AUTOMATED_PR) }}
```
If I remember correctly, it was because it resolved to `'true'` or `'false'` instead of `true` or `false`, which meant that as long it had any value it was always truthy (as a non-empty string), so I assumed it was a bug

dangrous commented:
Cool, thank you for the outline! I'll try to finish up a review by early next week. In the meantime, just summarizing what's _currently_ on the table:
* I think just one [comment](https://github.com/Expensify/App/pull/15560#discussion_r1181903510) from @AndrewGable still needs to be addressed
* Updating the repo directory removal
* A number of linting/other errors are now showing up in the GH checks (I think this might be partially due to the update of our ESLint config). Also signed commits, it looks like.

I'll keep on trucking through the code but other than that ^ it's looking good so far to me!

dangrous commented:
I haven't had a chance to go through more of the code, but thought I should check in on those other comments [above](https://github.com/Expensify/App/pull/15560#issuecomment-1536484406)... Let me know if you have any questions or if we can get that updated, and then go through for final checks!

radoslawkrzemien commented:
Hey @dangrous , sorry for the delay, I had to focus on another project and didn't have the time to look into these comments. I will try to action them before EoW

radoslawkrzemien committed:
Remove mock repo dir before tests

Added an utils method to force remove all remnants of the mock repo that might be left behind after forcibly terminating the tests previously. The removal happens before all tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Reorder README

Moved setup and running sections of the README up

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Rework logfiles creation

Logfiles are now created 1 per test, not per test suite, in a subdirectory named after the workflow under test

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
Hi @dangrous , I have added an utils function to `rm -rf` the mock repo directory before the whole test suite, in case the tests have been terminated prematurely without cleanup.
I have also changed the way the log files are generated, instead of `/logs/workflowName.log` they are now in `/logs/workflowName/testName.log`. Because there are now as many log files as there are tests they are grouped by the workflow to not get lost in them.
Lastly, I have rearranged the README file according to @AndrewGable 's comment

dangrous commented:
Cool! GH tests are still a bit unhappy with that latest version...

radoslawkrzemien committed:
Add `workflow_tests` to `ignorePatterns`

Workflow tests rely heavily on async functions and disallowing them in eslintrc is problematic

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix lint problems

Tons of minor changes to conform with eslint

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into 13604-local-testing-of-github-actions

radoslawkrzemien committed:
Fix failing tests

Resolved issues with tests that started failing after syncing with main repo

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Run prettier

Run prettier command to reformat the code

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Tiny reformat

To conform with eslint

See: https://github.com/Expensify/App/issues/13604

AndrewGable commented:
What's the latest on this?

dangrous commented:
I have not had time to take a look recently due to moderation; I can dig in next week.

Looks like we have a couple Travis errors (and conflicts) still, if you have time to look at those in the meantime @radoslawkrzemien that would be great!

dangrous commented:
That sounds like a good plan; let us know what you find out

radoslawkrzemien committed:
Job mocking

Added support for mocking entire jobs due to GH Actions reusable workflows not being able to be executed as steps

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Revert reusable workflow as step

Reverted change to cherryPick workflow using reusable workflow as step and adjusted tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Remove runs-on

Removed runs-on from cherryPick reusable workflow and allowed for setting it in the test

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Pass with to steps

When mocking a job, if the job has with clause it will be passed to all the mock steps

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Revert workflow change in finishReleaseCycle

Reverted the change in the finishReleaseCycle workflow reusable workflow to be executed again as a job and adjusted tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Revert workflow change in preDeploy

Reverted the change in the preDeploy workflow reusable workflow to be executed again as a job and adjusted tests

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into 13604-local-testing-of-github-actions

# Conflicts:
#	.github/workflows/platformDeploy.yml
#	.github/workflows/testBuild.yml
#	.gitignore
#	package-lock.json
#	package.json

radoslawkrzemien committed:
Resolve conflicts

Merged in main branch and resolved conflicts

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien commented:
@dangrous Sorry for another delay, I was OOO Thu-Tue.
I have added support for mocking entire jobs, reverted the changes to reusable workflows and adjusted the tests to mock them using the new feature. The tests are passing again and the GH Actions validation is passing as well

radoslawkrzemien committed:
Fix test

Fixed platformDeploy test which was failing after merge from main

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Fix test

Fixed testBuild test which was failing after merge from main

See: https://github.com/Expensify/App/issues/13604

radoslawkrzemien committed:
Merge remote-tracking branch 'origin/main' into 13604-local-testing-of-github-actions

# Conflicts:
#	.github/workflows/platformDeploy.yml
#	.github/workflows/testBuild.yml

radoslawkrzemien committed:
Fix tests

Fixed platformDeploy and testBuild tests which were failing after another merge from main

See: https://github.com/Expensify/App/issues/13604

dangrous commented:
Okay this is looking good! I'll keep going through the code itself but tests well, and I think (other than the new merge conflicts, we can ignore for now) the only *automated* issue here is the "Verify signed commits" check - not sure exactly how to fix that one.

Another, broader question as we go through, and this is one I should have asked earlier so I apologize:
* What's the use case for these? How will we use them day to day? This will help me understand how to look at them and make sure we're good to go

radoslawkrzemien commented:
I guess I need to do some rebasing and sign all the commits.

As for usage:
- I've added a command to package.json to trigger the tests with one command, so it can be run whenever there are changes made to the workflows to see if there are any functionalities that were unexpectedly broken. (manual check)
- Further, I could see potentially another GH Actions workflow that would run against every PR, check if there are changes in the workflows, and if there are to run that command automatically, blocking the PR from merging if there are workflow changes but tests don't pass (automatic CI check)
- I've added a bunch of tools and functions that allow for writing these tests in a similar fashion to regular Jest tests in JS/TS, so the learning curve is low for someone who already writes regular tests, plus features like file logs (since act-js doesn't print much for the user, but logs are verbose), etc.
- I've also added another command to package.json that autogenerates the test files stubs based on given GH Actions workflow file, which significantly speeds up writing the tests themselves
- Where I would see most use is in checking the branching paths in the workflows and conditions that steer the execution of the workflow, i.e. which jobs and which steps are/aren't executed in given conditions, which env vars and outputs are being set and with which values, etc.
- Less so for the concrete functionalities interacting with other services (like GitHub), these need to be either mocked, or provide the credentials to the test environment

Let me know if you have any more questions

dangrous commented on workflow_tests/README.md:
````diff
@@ -0,0 +1,422 @@
+```
+Assert results are as expected. This can, for example, include using `expect()` to check if the steps that should be executed have indeed been executed, steps that shouldn't run have not been executed, compare statuses (which steps succeeded, which failed) and step outputs. Outputs can include additional information, like input values, environmental variables, secrets (although these are usually  ...
+```javascript
+assertions.assertSomethingHappend(result);
````
```suggestion
assertions.assertSomethingHappened(result);
```

radoslawkrzemien replied:
Updated in https://github.com/Expensify/App/pull/21937

dangrous commented on workflow_tests/README.md:
```diff
@@ -0,0 +1,422 @@
+
+        // assert results (some steps can run in parallel to each other so the order is not assured
+        // therefore we can check which steps have been executed, but not the set job order
+        assertions.assertSomethingHappend(result);
```
```suggestion
        assertions.assertSomethingHappened(result);
```

radoslawkrzemien replied:
Updated in https://github.com/Expensify/App/pull/21937

dangrous commented on .github/workflows/warnCPLabel.yml:
```diff
@@ -19,7 +19,9 @@ jobs:
             If you applied the `CP Staging` label before the PR was merged, the PR will be be immediately deployed to staging even if the [open `StagingDeployCash` deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) is locked.
             However if you applied the `CP Staging` after the PR was merged it's possible it won't be CP'ed automatically. If you need it to be CP'ed to staging, tag a member of @Expensify/mobile-deployers to CP it manually, otherwise you can wait for it to go out with the next deploy.
-      - if: ${{ failure() }}
+
```
Do we need this blank line?

radoslawkrzemien replied:
Workflow was removed from `main` in the meantime so the question is a moot point now

dangrous commented on workflow_tests/assertions/preDeployAssertions.js:
```diff
@@ -0,0 +1,253 @@
+            {key: 'GITHUB_TOKEN', value: '***'},
+        ]),
+    ];
+    const osBotifyBody =
```
NAB - is there a way to get these from somewhere else, so that these tests won't break if we end up changing the copy?

radoslawkrzemien replied:
New assertion files can be generated using `npm run workflow-test:generate` command. So if at any point in time you change the workflow file and the assertions become outdated you can regenerate a new one based on the updated workflow with this command. It will however remove any changes added to the assertions file manually, so it would probably be best to keep a copy of the old file, generate new one and cherry pick what fits from both.
Or just copy the text from the new workflow

dangrous reviewed (changes requested):
Some small changes. Looking good overall! Would love your help @AndrewGable and @sobitneupane looking through this to make sure I'm not missing anything, as it's very large haha. 

With these changes, the (new) merge conflicts, and those signed commits, I think we'll be ready to go! Fingers crossed.

radoslawkrzemien commented:
Hi @dangrous . Signing 100+ already pushed commits, with 9k merged-in in the meantime from main turned out to be a bit more of a hussle than I had anticipated. I have decided to create a new branch, cherry-pick all the commits authored by me from this branch and sign them in the process. I will create a new PR from that new branch (draft [here](https://github.com/Expensify/App/pull/21937) ) and action all of the comments left here, as well as resolve conflicts

dangrous commented:
Sounds good! Let us know when it's ready.

radoslawkrzemien commented:
@dangrous , I have signed all the cherrypicked commits in the new branch, resolved conflicts with `main`, fixed the workflow tests that started to fail after changes in workflows done in the meantime and actioned the comments left in the discussion under this PR.
I have opened the new PR - https://github.com/Expensify/App/pull/21937 and I am closing this one.
LMK if something is unclear

radoslawkrzemien commented:
Actually, there's one more workflow that has been modified, I'm on it
